### ISSUE TITLE
[Solved] Upgrade to xUbuntu 24.04 broke tuxedo_io

### ISSUE THREAD

**Issue by Vitamin-X** [tuxedocomputers/tuxedo-control-center#402]
Upgraded to xUbuntu 24.04 yesterday and after a reboot nothing happened, the kernel wont start...
After a bunch of package re installs inside a chroot environment, i notice that the tuxedo repository was missing and it de installed some packages, after re-adding it and installing the tuxedo stuff and the NVIDIA driver I'm able to boot again, but now I'm getting an "tuxedo_io" error and I'm unable to control the fan speed

`cat /var/log/dmesg | grep "tcc\|tux"
[    5.361203] kernel: tuxedo_keyboard: loading out-of-tree module taints kernel.
[    5.361206] kernel: tuxedo_keyboard: module verification failed: signature and/or required key missing - tainting kernel
[    5.361580] kernel: tuxedo_keyboard: module init
[    5.362299] kernel: tuxedo_io: disagrees about version of symbol uniwill_get_device_features
[    5.362301] kernel: tuxedo_io: Unknown symbol uniwill_get_device_features (err -22)
[    5.462192] kernel: tuxedo_io: disagrees about version of symbol uniwill_get_device_features
[    5.462194] kernel: tuxedo_io: Unknown symbol uniwill_get_device_features (err -22)
[    5.490251] kernel: tuxedo_keyboard: Set keyboard enabled to: 1
[    5.494639] kernel: tuxedo_keyboard: Set keyboard backlight mode on CUSTOM
[    5.504900] kernel: input: TUXEDO Keyboard as /devices/platform/tuxedo_keyboard/input/input15
[    5.511661] kernel: tuxedo_io: disagrees about version of symbol uniwill_get_device_features
[    5.511664] kernel: tuxedo_io: Unknown symbol uniwill_get_device_features (err -22)
[    5.553483] kernel: tuxedo_io: disagrees about version of symbol uniwill_get_device_features
[    5.553487] kernel: tuxedo_io: Unknown symbol uniwill_get_device_features (err -22)
[    5.580791] kernel: tuxedo_io: disagrees about version of symbol uniwill_get_device_features
[    5.580795] kernel: tuxedo_io: Unknown symbol uniwill_get_device_features (err -22)
[    5.595696] kernel: tuxedo_io: disagrees about version of symbol uniwill_get_device_features
[    5.595698] kernel: tuxedo_io: Unknown symbol uniwill_get_device_features (err -22)
[    5.913372] kernel: tuxedo_io: disagrees about version of symbol uniwill_get_device_features
[    5.913375] kernel: tuxedo_io: Unknown symbol uniwill_get_device_features (err -22)
[    5.957476] kernel: intel_tcc_cooling: Programmable TCC Offset detected
[    6.395609] kernel: tuxedo_io: disagrees about version of symbol uniwill_get_device_features
[    6.395611] kernel: tuxedo_io: Unknown symbol uniwill_get_device_features (err -22)`

`dkms status
acpi-call/1.2.2, 6.5.0-25-generic, x86_64: installed
acpi-call/1.2.2, 6.5.0-26-generic, x86_64: installed
acpi-call/1.2.2, 6.5.0-27-generic, x86_64: installed
acpi-call/1.2.2, 6.5.0-28-generic, x86_64: installed
acpi-call/1.2.2, 6.8.0-31-generic, x86_64: installed
acpi-call/1.2.2, 6.8.0-32-generic, x86_64: installed
acpi-call/1.2.2, 6.8.0-35-generic, x86_64: installed
acpi-call/1.2.2, 6.8.0-36-generic, x86_64: installed
acpi-call/1.2.2, 6.8.0-38-generic, x86_64: installed
nvidia/550.67, 6.5.0-28-generic, x86_64: installed
nvidia/550.67, 6.8.0-36-generic, x86_64: installed
nvidia/550.67, 6.8.0-38-generic, x86_64: installed
tuxedo-drivers/4.5.2, 6.5.0-28-generic, x86_64: installed (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!)
tuxedo-drivers/4.5.2, 6.8.0-31-generic, x86_64: installed
tuxedo-drivers/4.5.2, 6.8.0-32-generic, x86_64: installed (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!)
tuxedo-drivers/4.5.2, 6.8.0-35-generic, x86_64: installed (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!)
tuxedo-drivers/4.5.2, 6.8.0-36-generic, x86_64: installed (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!)
tuxedo-drivers/4.5.2, 6.8.0-38-generic, x86_64: installed (WARNING! Diff between built and installed module!)
tuxedo-keyboard/3.2.14, 6.5.0-21-generic, x86_64: installed
tuxedo-keyboard/3.2.14, 6.5.0-25-generic, x86_64: built
tuxedo-keyboard/3.2.14, 6.5.0-26-generic, x86_64: installed (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!)
tuxedo-keyboard/3.2.14, 6.5.0-27-generic, x86_64: installed (WARNING! Diff between built and installed module!)
tuxedo-keyboard/3.2.14, 6.5.0-28-generic, x86_64: installed
tuxedo-keyboard/3.2.14, 6.8.0-31-generic, x86_64: installed (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!)
tuxedo-keyboard/3.2.14, 6.8.0-32-generic, x86_64: installed (WARNING! Diff between built and installed module!)
tuxedo-keyboard/3.2.14, 6.8.0-35-generic, x86_64: installed (WARNING! Diff between built and installed module!)
tuxedo-keyboard/3.2.14, 6.8.0-36-generic, x86_64: installed (WARNING! Diff between built and installed module!)
tuxedo-keyboard/3.2.14, 6.8.0-38-generic, x86_64: installed (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!)`

when i reinstall the "tuxedo-driver" with 
`dkms remove -m tuxedo-drivers -v 4.5.2 -k 6.8.0-36-generic && dkms install -m tuxedo-drivers -v 4.5.2  -k 6.8.0-36-generic`
or "tuxedo-keyboard"
`dkms remove -m tuxedo-keyboard -v 3.2.14 -k 6.8.0-36-generic && dkms install -m tuxedo-keyboard -v 3.2.14 -k 6.8.0-36-generic`

the message of "(WARNING! Diff between built and installed module!)" sometimes moves around, either way it wont change the behavior.

`tuxedo-drivers/4.5.2, 6.8.0-36-generic, x86_64: installed (WARNING! Diff between built and installed module!)
tuxedo-keyboard/3.2.14, 6.8.0-36-generic, x86_64: installed (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!)`

so far kernel 6.8.0-36 is the only one able to start, didn't had any luck with the other versions yet.

also Tried to purge tuxedo-keyboard to reinstall it, also compiled a version from the source code, but no luck either.

compiling tuxedo_keyboard also throws 
`Skipping BTF generation for ./tuxedo-keyboard/./src/tuxedo_keyboard.ko due to unavailability of vmlinux
Skipping BTF generation for ./tuxedo-keyboard/./src/clevo_wmi.ko due to unavailability of vmlinux
Skipping BTF generation for ./tuxedo-keyboard/./src/clevo_acpi.ko due to unavailability of vmlinux
Skipping BTF generation for ./tuxedo-keyboard/./src/tuxedo_io/tuxedo_io.ko due to unavailability of vmlinux
Skipping BTF generation for ./tuxedo-keyboard/./src/uniwill_wmi.ko due to unavailability of vmlinux`

after
modprobe tuxedo-keyboard #works
modprobe tuxedo_io #doesn't
`modprobe: ERROR: could not insert 'tuxedo_io': Invalid argument`

But the keyboard+lights and touch pad are working just the fans aren't shown inside the software
![grafik](https://github.com/tuxedocomputers/tuxedo-control-center/assets/8036752/6cc9b331-d32b-4c2c-ab69-6bdd6c80c5e9)

but now im out of ideas, already spent the whole day on it.
Anyone got an idea except reinstalling the whole System?
that here is now transformed to a log.txt on my desktop to keep track on what happened

now trying "apt purge tuxedo-drivers tuxedo-keyboard tuxedo-control-center && apt install tuxedo-drivers tuxedo-keyboard tuxedo-control-center"
maybe the recompiled debs give me any luck.

keyboard lights turned off while doing so and turned on again after.

while installing 
`tuxedo_io.ko.zst:
Running module version sanity check.
Error! Module version 0.3.9 for tuxedo_io.ko.zst
is not newer than what is already found in kernel 6.8.0-36-generic (0.3.9).
You may override by specifying --force.`
popped up, is that important? 
dkms cant even see that file...

i just removed the  folder "/lib/modules/6.8.0-36-generic/updates/src/tuxedo_io"  so lets see if apt can install it now...

OMG it freaking worked!!!!!

TLDR

so way to solve it on kernel 6.8.0-36:
(it will remove all your tuxedo settings)
`rm -R /lib/modules/6.8.0-36-generic/updates/src/tuxedo_io
apt purge tuxedo-drivers tuxedo-keyboard tuxedo-control-center && apt install tuxedo-drivers tuxedo-keyboard tuxedo-control-center`

kinda hope google will find that here cuz i had absolutely no luck finding anything even close so im posting it anyway

gonna try tomorrow to fix the other kernels, so far i get enough for today.

Wasted hours-> 14:27

haven't rebooted yet but 
`dkms status
[...]
tuxedo-drivers/4.5.3, 6.8.0-36-generic, x86_64: installed
tuxedo-keyboard/3.2.14, 6.8.0-36-generic, x86_64: installed (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!) (WARNING! Diff between built and installed module!)
[...]`
the keyboard still shows an error but the tuxedo-drivers doesn't now

**Comment by juergh** [comment]
Tuxedo dist-upgrades don't seem to work yet. You probably have to reenable the tuxedo repo afterwards. From the log above you're running the stock Ubuntu kernel, not the tuxedo kernel. Try installing that: `apt install linux-tuxedo-24.04`. Also note that the tuxedo version of linux-firmware is behind Ubuntu's so you won't get whatever tuxedo linux-fiwmare is supposed to provide.

**Comment by Vitamin-X** [comment]
i didnt knew there was such package, but i tryed installling the tuxedo kernel itself at some point to, but no luck, for now it seems that the  tuxedo_io module broke and wiping it fixed everything

but yea the repo was gone, so i re added the newest one with...
`echo "deb https://deb.tuxedocomputers.com/ubuntu noble main" > /etc/apt/sources.list.d/tuxedo.list`

looking back at the logs it said "You may override by specifying --force." when i tried to install the packages, maybe the "--force" option is something they might wanna add to the installer? so it will prevent such things from happening again, 

otherwise also no clue why it happen in the first place, maybe the compiler exited for some reason

anyhow at the end i just wanted to post it somewhere so google could pick it up and help the next person cuz i didn't had any luck with it and they might dont have the time or experience to solve it

sooo Hello next person 👋
you probably need to change the kernel versions above to match yours,
you can look them up with "uname -r"
